### DISCUSSION TITLE
`RETURNING` for SQLite

### DISCUSSION THREAD

**Discussion by Boscop** [diesel-rs/diesel#2684] [Ideas]
SQLite now supports `RETURNING`: http://antonz.org/sqlite-3-35/
https://sqlite.org/releaselog/3_35_0.html

It would be great if diesel could support it as well :)

---
> What do you want to do, and how do you expect Diesel to support you with that?

Currently, I'm doing things like
```rust
diesel::insert_into(records).values(new).execute(conn)?;
let r = records.find(sql("last_insert_rowid()")).get_result(conn)?;
```
I'd prefer to be able to do:
```rust
let r = diesel::insert_into(records).values(new).returning(records::all_columns).get_result(conn)?;
```

**Comment by Talden** [comment]
It looks support for sqlite-3.35 merged in https://github.com/rusqlite/rusqlite/pull/917 in the last 24 hours.

No idea when that might reach a libsqlite3-sys release or what it will take for diesel to be in a position to take such a release.

The CTE and RETURNING changes both have the potential for measurable improvements.  DROP COLUMN is a nice quality-of-life change too.

**Reply by weiznich** [reply]
That's quite likely not really related for diesel, as diesel uses the ffi interface and non of this features require any changed function there.

**Comment by weiznich** [comment]
According to our contributions guide lines this is a feature request. As such it does not belong to our issue tracker, but should be filled as a proposal in our discussion forum. I've transferred it there now, but I expect that you add the necessary information as specified [here](https://github.com/diesel-rs/diesel/blob/master/CONTRIBUTING.md#submitting-feature-requests).

Otherwise [this](https://github.com/diesel-rs/diesel/issues/771#issuecomment-798431770) comment applies here as well.

**Comment by FelixHenninger** [comment]
Dear diesel team, hi everyone,

thanks for pointing me to this discussion! (and sorry for resurrecting an old issue in the meantime) I'm also very interested in this feature, and would be thrilled to get it to work. I'm not the original poster, but here's what I would answer to the feature request prompts:

> What do you want to do, and how do you expect Diesel to support you with that?

I would love to use the new `RETURNING` clause supported by SQLite 3.35.0 and later. This feature, which is already supported by Diesel for Postgres, would need to be enabled for SQLite also.

> How might this be added to Diesel?

Because the implementation is designed to match Postgres, the logic that is already in place should ideally apply directly. I have started [trying to make the necessary changes on a fork](https://github.com/diesel-rs/diesel/compare/master...FelixHenninger:master).

> What are possible alternatives?

A [multitude of workarounds](https://github.com/diesel-rs/diesel/issues/771) for this issue have already been put forward that would continue to partially work (failure scenarios remain) -- admittedly this is a quality-of-life issue that would vastly simplify my own code, but would not be strictly necessary. The main alternative I see is to wait for a while.

> Are there any disadvantages?

I'm not nearly informed enough to answer this question, but I can take a couple guesses:
* Support for `RETURNING` is a fairly recent change to SQLite, and not likely to be deployed widely. Thus, there probably needs to be some way to handle older versions without this feature. This could be through code, or via documentation.
* The changes I propose above [don't pass CI](https://github.com/FelixHenninger/diesel/runs/2222102836?check_suite_focus=true#step:30:15), which might point to a larger issue that needs to be resolved first (I tried -- the tests appear straightforward to me, and I would need help to figure out what's going on).

Alright, this is just my 2ct -- I hope this helps, and would be thrilled to help bring this feature to life. Any hints, especially regarding the last issues, would be greatly appreciated -- I'm happy to try to tackle both, but would need direction (I've joined the Gitter room in case anybody would like to get in touch). Thanks again, everyone!

**Comment by weiznich** [comment]
As I consider this as an feature that would be nice to have let's write down a list of open questions here first:

# How do we handle the sqlite version requirement given that this is only available for >=3.35

1. Drop support for sqlite versions < 3.35. -> I do not consider that to be really an option here
2. Put this feature behind a feature flag that is disabled by default. 
3. Build time detection of the linked sqlite so that we only include the corresponding code if the sqlite version is 3.35 or newer. This will likely break as soon as someone dynamically links sqlite and then uses a older version afterwards
4. Provide a runtime fallback for returning clauses issuing a second query based on the sqlite version. That would likely require larger changes to our query builder to enable this kind of dynamic query building based on connection properties. Also this opens the question how a generic fall back would look like.
5. Provide extensive documentation that this feature is only available starting from sqlite version 3.35, like we've already do this for upsert. In my opinion that's not really an option here as this feature is much more prominently placed then the upsert functionality.

Input on the different options is welcome.

# Implementation

Implementing this feature is as simple as adding an impl of `SupportsReturningClause` for `Sqlite`. The real work here is to fix the CI, write some documentation for this feature (especially that it won't support batch inserts for now) and enable at least a few tests. Additionally we need to decide how to solve the sqlite version problem.

**Reply by FelixHenninger** [reply]
Hi everyone,

thanks a lot for outlining the critical questions! Not being a team member I've been hesitant to comment on this, so please let me know if I'm overstepping here -- but in the interest of starting a discussion, my strong (naive) preference would be for option two, the feature flag, at least for now, plus documentation. I would argue (and agree with your reasoning above) that this is the safest option for the time being, and does not preclude the other choices which you might want to adopt at a later point once more recent SQLite versions become more widespread.

I'm happy to add the flag to [my branch](https://github.com/felixhenninger/diesel/), and do the requisite CI work -- would you want the flag to be a dimension in your testing matrix? (my hunch is that you might want a compile test to fail without the flag enabled?)

As discussed on Gitter, the `RETURNING` clause does not work for bulk inserts -- I haven't understood why, but would be happy to help investigate.

Kind regards, and many thanks,


-Felix

**Reply by weiznich** [reply]
> Not being a team member I've been hesitant to comment on this, so please let me know if I'm overstepping here

Discussions on feature requests are open for anyone. Their main purpose is to collect feedback from our users, independly if they are part of the diesel maintainer team or not. That means it's great to add comments here :+1:

> As discussed on Gitter, the RETURNING clause does not work for bulk inserts -- I haven't understood why, but would be happy to help investigate.

Sqlite does support bulk inserts in theory. In practice does the implementation of inserts in diesel assume that there is some way to signal the database that it should use the default value for some column. That is something that is not supported by SQLite. We work around this by just skipping the corresponding column from the insert column list for single item insert statement, as SQLite then falls back to the default value. That's not possible for bulk insert statements, as each inserted row could miss a other column. To work around this we've opted for just running separate insert statements for each row. (See [here](https://github.com/diesel-rs/diesel/blob/3f3ce57ca1a89274ecf915b70826234c45bd261e/diesel/src/query_builder/insert_statement/mod.rs#L290-L309) for the corresponding code). Now that's something that works, but it's not that efficient, which means we would likely accept an alternative implementation there as long as it works for all existing cases. Unfortunately is that something where I do not have a clear idea how to implement that, therefore that's the current state there.

> I'm happy to add the flag to my branch, and do the requisite CI work -- would you want the flag to be a dimension in your testing matrix? (my hunch is that you might want a compile test to fail without the flag enabled?)

I would be fine with having this impl behind a feature flag for now. This should correctly be checked by the compile_fail tests + have a changelog entry. Otherwise I think it's a probably good idea to "upgrade" the "sqlite-bundled" CI job to run the complete test suite and check there if this feature work. As the "sqlite-bundled" feature embeds the latest sqlite version it should just work there without having to build everything manually from source.

**Reply by FelixHenninger** [reply]
Thanks a lot for your explanations, and for your kind words! I would be thrilled to contribute; apologies for the fuss I'm causing!

> Sqlite does support bulk inserts in theory. In practice does the implementation of inserts in diesel assume that there is some way to signal the database that it should use the default value for some column. That is something that is not supported by SQLite. We work around this by just skipping the corresponding column from the insert column list for single item insert statement, as SQLite then falls back to the default value. That's not possible for bulk insert statements, as each inserted row could miss a other column. To work around this we've opted for just running separate insert statements for each row.

Ok, this makes total sense; what I'm wondering about is whether a proper implementation of `RETURNING` would need to change the workaround. Right now, it returns the number of inserts (wrapped in a `QueryResult`), but no further data as far as I can tell. Would it be difficult to instead collate the individual records and return the batch?

If I'm not mistaken, the tests that are failing are the ones _without_ default values (`RETURNING` is not tested in combination with `DEFAULT` as far as I can tell), so my understanding is that the issue is not contingent on a default being set -- I haven't fully understood how the specific error occurs that complains about a lack of `SupportsDefaultKeyword` on SQLite.

> I would be fine with having this impl behind a feature flag for now. This should correctly be checked by the compile_fail tests + have a changelog entry.

That's great, thanks! I've been thinking about how to test this correctly -- the way I imagine it, you would want to see the tests both fail without the flag and succeed with it, and that would require a separate run of the `compile_tests` with it's own `Cargo.toml` to set the feature flag, or a separate CI task to run `cargo` with the appropriate CLI flags. Is that the right idea, or would you be happy only compile-testing the default flag set (in which case all tests are already in place)?

**Reply by weiznich** [reply]
> I would be thrilled to contribute; apologies for the fuss I'm causing!

Asking well written questions is always welcome, so no reason to apologies for anything :wink:.

> If I'm not mistaken, the tests that are failing are the ones without default values (RETURNING is not tested in combination with DEFAULT as far as I can tell), so my understanding is that the issue is not contingent on a default being set -- I haven't fully understood how the specific error occurs that complains about a lack of SupportsDefaultKeyword on SQLite.

The point is not that the test cases itself use the default keyword, but that the only implementation for [insert values for all backends](https://github.com/diesel-rs/diesel/blob/fe989447932a6c6674d0f781bd5e715d21518d73/diesel/src/insertable.rs#L130-L135). Every insertable expression like `some_column.eq("foo")` or what `#[derive(Insertable)]` generates internally gets compiled down to this enum. The `Default` value there can be easily constructed via a `None` value in any field of your `#[derive(Insertable)]` struct, so just dropping support for this feature is not possible. As already mentioned this is no real problem for single value inserts as we can simply drop the columns with default values here. However it is problematic for batch insert values. Therefore we limit the `QueryFragment`  impl (which provides a way to translate a typed query ast into a query) for `InsertStatements` [here](https://github.com/diesel-rs/diesel/blob/fe989447932a6c6674d0f781bd5e715d21518d73/diesel/src/query_builder/insert_statement/mod.rs#L183). `CanInsertInSingleQuery` is restricted to backends that support the default keyword [here](https://github.com/diesel-rs/diesel/blob/fe989447932a6c6674d0f781bd5e715d21518d73/diesel/src/insertable.rs#L99-L115) for the corresponding batch insert containers. 
We could probably work around this for now by adding `LoadQuery` impls similar to the existing `ExecuteDsl` impls [here](https://github.com/diesel-rs/diesel/blob/3f3ce57ca1a89274ecf915b70826234c45bd261e/diesel/src/query_builder/insert_statement/mod.rs#L290-L309). 

>  I've been thinking about how to test this correctly -- the way I imagine it, you would want to see the tests both fail without the flag and succeed with it, and that would require a separate run of the compile_tests with it's own Cargo.toml to set the feature flag, or a separate CI task to run cargo with the appropriate CLI flags. Is that the right idea, or would you be happy only compile-testing the default flag set (in which case all tests are already in place)?

I would prefer to have a single separate CI job that tests the corresponding implementation. Maybe we could upgrade the bundled-sqlite test run to do this. We would need to run tests for `diesel`, `diesel_compile_tests` and `diesel_tests` there. Those crates would need corresponding feature flags.

**Reply by pickfire** [reply]
Do we want to do feature versioning mechanism?
```toml
[features]
...
sqlite_3_35_0 = ["sqlite_3_34_0"]
sqlite = ["sqlite_3_35_0"]
```
So we can put like `#[cfg(feature = "sqlite_3_35_0")]`?

**Reply by weiznich** [reply]
We do not want to have a separate feature flag per released sqlite version, but only one feature flag for this specific implementation. I would likely prefer to name in such a way that it is clear that this feature flag is related to returning clauses and requires sqlite 3.35 or newer. So maybe just `returning_clauses_for_sqlite_3_35` or something like that. Also the `sqlite` feature flag should **not** imply this feature, because otherwise it would be impossible to use the sqlite backend without using this feature.

**Reply by pickfire** [reply]
We could do `returning_clauses_for_sqlite_3_35`, but later when more features come up as new sql versions come up, we have a lot more feature flags which will blow the complexity. Whereas settings the feature flag only by version makes stuff easy. Checking just as single version is easier than checking for all the features you need (and it won't even show up in offline docs if you didn't enable the feature). So doing just feature flags by version rather than sqlite specific feature would be easier.

Let me give an example, if you use rust, do you want to set a rust version or you want to opt in to features like async, match, matches, question_mark, pub_crate, path, ...? If we do it using features, it would be weird if someone toggle a feature that is only available in 3.35 but not another feature available in 3.35.

**Reply by EyeOfPython** [reply]
Any progress on this?

**Reply by weiznich** [reply]
@EyeOfPython We are happy to receive a contribution for example from you for this feature.

**Reply by toxeus** [reply]
I've opened a PR to address this: https://github.com/diesel-rs/diesel/pull/3004
